### PR TITLE
Interval slicer issue

### PR DESCRIPTION
**Pull request by kawindme**

### CODE REVIEW

File: manual_tests/manual_test_intervals.py [hunk 1]
```diff
@@ -202,8 +202,8 @@
 
 # reference="left", right_open,
 ref_references = [1, 2, 3, 4, 5]
-ref_intervals = [[1.2], [1.5, 2.4], [2.5, 2.6, 3.1], [3.5, 3.6, 4.0], [5.0]]
-ref_boundaries = [[0.5, 1.5], [1.5, 2.5], [2.5, 3.5], [3.5, 4.5], [4.5, 5.5]]
+ref_intervals = [[1.2, 1.5], [2.4, 2.5, 2.6], [3.1, 3.5, 3.6], [4.0], [5.0]]
```
Comment (ahaselsteiner) [inline manual_tests/manual_test_intervals.py:205]: Is this datastructure correct? ref_intervals = [[1.2, 1.5], [2.4, 2.5, 2.6], [3.1, 3.5, 3.6], [4.0], [5.0]]
What does it describe? If these are the reference values for intervals why is not one value per interval? E.g. the middle of the interval?

Probably I should know that as we probably together discussed it (?) but at the moment it does not make any sense to me.

File: manual_tests/manual_test_intervals.py [hunk 2]
```diff
@@ -216,9 +216,9 @@
 
 
 # reference="left", left_open
-ref_references = [1, 2, 3, 4, 5]
-ref_intervals = [[1.2, 1.5], [2.4, 2.5], [2.6, 3.1, 3.5], [3.6, 4.0], [5.0]]
-ref_boundaries = [[0.5, 1.5], [1.5, 2.5], [2.5, 3.5], [3.5, 4.5], [4.5, 5.5]]
+ref_references = [1, 2, 3, 4]
+ref_intervals = [[1.2, 1.5], [2.4, 2.5, 2.6], [3.1, 3.5, 3.6, 4.0], [5.0]]
```
Comment (ahaselsteiner) [inline manual_tests/manual_test_intervals.py:220]: See above.

File: tests/test_intervals.py [hunk 3]
```diff
@@ -131,6 +156,21 @@ def woi_params_and_ref(request):
     ]
 )
 def noi_params_and_ref(request):
+    """
+    Fixture that applies the NumberhOfIntervalsSlicer to the prepared test case.
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:160]: Typo "NumberOf"

File: virocon/intervals.py [hunk 4]
```diff
@@ -158,14 +160,31 @@ def _slice(self, data):
                 data_max = np.max(data)
 
         width = self.width
-        interval_references = np.arange(data_min, data_max + width, width)
+        interval_references = np.arange(data_min, data_max + width, width)+ 0.5*width
```
Comment (ahaselsteiner) [inline virocon/intervals.py:163]: space before "+" is missing. Generally, have you run black on the module?

File: tests/test_intervals.py [hunk 5]
```diff
@@ -11,6 +11,16 @@
...
+    Function to create test data.
+
+    Returns
+    -------
+    Numpy array
+        Test data.
+
+    """
+    
     return np.array([1.2, 1.5, 2.4, 2.5, 2.6, 3.1, 3.5, 3.6, 4.0, 5.0])
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:24]: These numbers look the same as the ones used for "ref_intervals" (which I do not understand).

**Review by ahaselsteiner — COMMENTED** [review]
Thanks! While I did not evaluate whether the new functionality works correctly (I hope this is taken care of with the tests), I did not understand some lines in the tests. The variable "ref_intervals" was unclear to me (see individual comments). What does ref_intervals represent?

File: manual_tests/manual_test_intervals.py (hunk 1 above)
Comment (lenastroeer) [reply manual_tests/manual_test_intervals.py:205]: The datapoint are sorted into ref_intervals (the naming of the variables are confusing). The references of the intervals are indicated in re_references.

Comment (lenastroeer) [reply manual_tests/manual_test_intervals.py:205]: @kawindme should we delete this manual test?

File: tests/test_intervals.py (hunk 5 above)
Comment (lenastroeer) [reply tests/test_intervals.py:24]: These are test data which should be sorted into the intervals.

File: tests/test_intervals.py [hunk 6]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    my_slices, my_references, my_boundaries = width_slicer.slice_(test_data)
+    my_intervals = [test_data[slice_] for slice_ in my_slices]
+    np.testing.assert_array_equal(my_references, ref_references)
+    assert len(my_intervals) == len(ref_intervals)
+    for i in range(len(my_intervals)):
+        np.testing.assert_array_equal(my_intervals[i], ref_intervals[i])
+    for i in range(len(my_boundaries)):
+        np.testing.assert_array_equal(my_boundaries[i], ref_boundaries[i])
+
+
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:937]: Delete empty lines

File: tests/test_intervals.py [hunk 7]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # n_points= 2
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:370]: What does this comment mean? Please either describe it clearly using a full sentence or delete it.

File: tests/test_intervals.py [hunk 8]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # n_points= 3, last_full=True
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:401]: See above.

File: tests/test_intervals.py [hunk 9]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # n_points= 3, last_full=False
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:432]: See above.

File: tests/test_intervals.py [hunk 10]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # n_points= 3, last_full=True, min_n_points=3
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:463]: See above.

File: tests/test_intervals.py [hunk 11]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # n_points= 3, last_full=True, reference=np.mean
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:494]: See above.

File: tests/test_intervals.py [hunk 12]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # n=2 include_max=True
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:525]: See above.

File: tests/test_intervals.py [hunk 13]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # n=2, include_max=False
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:555]: See above.

File: tests/test_intervals.py [hunk 14]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """ 
+    
+    # n=3 include_max=True
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:586]: See above.

File: tests/test_intervals.py [hunk 15]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """ 
+        
+    # n=3 include_max=True, min_n_points=3, min_n_intervals=2
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:618]: See above.

File: tests/test_intervals.py [hunk 16]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """ 
+        
+    # n=3 include_max=False
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:650]: See above.

File: tests/test_intervals.py [hunk 17]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """ 
+        
+    # n=2 include_max=True, value_range=(0, 5)
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:681]: See above.

File: tests/test_intervals.py [hunk 18]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """ 
+    
+    # n=2 include_max=True, reference=np.median
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:712]: See above.

File: tests/test_intervals.py [hunk 19]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """ 
+    
+    # n=3 include_max=True, min_n_points=3, min_n_intervals=3
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:743]: See above.

File: tests/test_intervals.py [hunk 20]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    ref_intervals = [[1.2, 1.5, 2.4], [2.5, 2.6, 3.1, 3.5, 3.6]]
+    ref_width = (5 - 1.2) / 3
+    ref_references = [1.2 + ref_width / 2, 1.2 + 3 * ref_width / 2]
+    ref_boundaries = [(c - ref_width / 2, c + ref_width / 2) for c in ref_references]
+    number_slicer = NumberOfIntervalsSlicer(3, min_n_points=3, min_n_intervals=3)
+    try:
+        my_slices, my_references, my_boundaries = number_slicer.slice_(test_data)
+    except RuntimeError:
+        pass  # we expect a RuntimeError if there are less than min_n_intervals intervals
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:752]: Please start the comment with a capital letter and end it with a dot (full sentence).

File: tests/test_intervals.py [hunk 21]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """
+    
+    # reference="left", right_open,
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:771]: See above.

File: tests/test_intervals.py [hunk 22]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    ----------
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """   
+    # reference="left", left_open
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:801]: See above.

File: tests/test_intervals.py [hunk 23]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    ----------
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """    
+    # reference="center", right_open
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:833]: See above.

File: tests/test_intervals.py [hunk 24]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    ----------
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """    
+    # reference="center", left_open
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:863]: See above.

File: tests/test_intervals.py [hunk 25]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    ----------
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """    
+    # reference=median, right_open
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:893]: See above.

File: tests/test_intervals.py [hunk 26]
```diff
@@ -349,3 +349,615 @@ def test_points_per_interval_slicer(test_data, ppi_params_and_ref):
...
+    ----------
+    test_data : numpy array
+        Data to test the interval slicer.
+ 
+    Returns
+    -------
+    None.
+
+    """    
+    # reference="left", right_open, min_n_points=2
```
Comment (ahaselsteiner) [inline tests/test_intervals.py:923]: See above.

**Review by ahaselsteiner — COMMENTED** [review]
@lenastroeer Thanks! Looks good - I just found a couple of things, which are about code formatting / commenting that should be fixed before merging (there is also one comment about a typo still open from my last review).

**Review by ahaselsteiner — APPROVED** [review]

**Pull request merged**